File: app/lib/dtos/database/activities/comments.dart
// Package imports:
import 'package:freezed_annotation/freezed_annotation.dart';

// Project imports:
import 'package:app/dtos/database/common/fl_meta.dart';
import 'package:app/dtos/database/common/media.dart';
import 'mentions.dart';

part 'comments.freezed.dart';
part 'comments.g.dart';

@freezed
class Comment with _$Comment {
  const factory Comment({
    @JsonKey(name: '_fl_meta_') FlMeta? flMeta,
    @Default('') String data,
    @Default('') @JsonKey(name: 'reaction_id') String reactionId,
    @Default('') @JsonKey(name: 'activity_id') String activityId,
    @Default('') @JsonKey(name: 'user_id') userId,
    @Default([]) @JsonKey(fromJson: Mention.fromJsonList, toJson: Mention.toJsonList) List<Mention> mentions,
    @Default([]) @JsonKey(fromJson: Media.fromJsonList, toJson: Media.toJsonList) List<Media> media,
  }) = _Comment;

  factory Comment.fromJson(Map<String, dynamic> json) => _$CommentFromJson(json);
}

File: app/functions/src/dto/comments.ts
import { MediaJSON } from "./media";
import { MentionJSON } from "./mentions";
import { FlMeta, FlMetaJSON } from "./meta";

export const commentSchemaKey = "comments";

export interface CommentJSON {
    _fl_meta_?: FlMetaJSON;
    data?: string;
    reaction_id?: string;
    activity_id?: string;
    user_id?: string;
    origin?: string;
    mentions?: MentionJSON[];
    media?: MediaJSON[];
}

export class Comment {
    _fl_meta_?: FlMeta;
    data?: string;
    reaction_id?: string;
    activity_id?: string;
    user_id?: string;
    origin?: string;
    mentions?: MentionJSON[];
    media?: MediaJSON[];

    constructor(json: CommentJSON) {
        this._fl_meta_ = json._fl_meta_ ? new FlMeta(json._fl_meta_) : undefined;
        this.data = json.data;
        this.reaction_id = json.reaction_id;
        this.activity_id = json.activity_id;
        this.user_id = json.user_id;
        this.origin = json.origin;
        this.mentions = json.mentions;
        this.media = json.media;
    }

    public toJSON(): CommentJSON {
        return {
            _fl_meta_: this._fl_meta_ ? this._fl_meta_.toJSON() : undefined,
            data: this.data,
            reaction_id: this.reaction_id,
            activity_id: this.activity_id,
            user_id: this.user_id,
            origin: this.origin,
            mentions: this.mentions,
            media: this.media,
        };
    }

    static FromJSONList(jsonList: CommentJSON[]): Comment[] {
        const list: Comment[] = [];
        if (jsonList) {
            jsonList.forEach((json) => {
                list.push(new Comment(json));
            });
        }

        return list;
    }
}